app/schemas.py

```diff
@@ -23,21 +21,21 @@
 # See: https://pydantic-docs.helpmanual.io/usage/types/#datetime-types for field types
 from pydantic import BaseModel
 
-
 # Local application imports
 # -------------------------
 # None.
-#
-#
-# Schemas?
-# ========
+
+
+# Schemas
+# =======
 class LogItemIncoming(BaseModel):
     """
     This class defines the schema for what we can expect to get from a logging event.
     Because we are using pydantic type verification happens automatically, if we want
     to add additional constraints we can do so.
     """
 
+    # :index:`TODO`: Is there any way to specify max string lengths? The database has fixed-length fields for some of these.
```

Comment:
Schemas contains Pydantic models.  There is a plugin for FastAPI to allow one to generate Pydantic schemas from Sqlalchemy models, I think we ought to investigate using that.  Otherwise there is some inevitable duplication of work.  But I do really like using these schemas and probably have a lot more to learn about them.

Reply:
I agree. When I first started on the NewServer (Flask-based) several years ago, I spent a lot of time writing validation code, something sorely lacking in web2py. Some helpful ideas from there:

- Define a standard `id` column type and re-use it in all tables.
- I've created a Boolean data type that works with web2py/DAL's weird boolean.
- Standard columns for all answer tables we define once and re-use.
- Some [docs](https://github.com/bjones1/NewServer/blob/master/runestone/api/endpoints.py#L125) on the `hsblog` endpoint. Does this look right? If so, we should incorporate it in the current code.

Reply:
Good ideas except for having a standard numeric id.  I honestly cringe every time I have to deal with that.  WE HAVE good logical primary keys for many of our tables and it hurts to have to use an arbitrary number rather than something meaningful just because of web2py.  Any good class on data modeling (which I have taught many times) will have you always prefer good natural primary keys over an arbitrary number.

Reply:
Can you integrate your Boolean data type for use with our models?  That seems like a wonderful solution.

Reply:
I'm happy to follow your lead on avoiding an ID. All my knowledge of databases is self-taught; I've never had any formal instruction. However, I think we need to select a more consistent naming convention -- for example, `course_id` is a string referring to the course name in the `useinfo` table, while it's an int referring to the id in the `code` table. This sort of nomenclature confusion makes it hard to write understandable code.

Perhaps we can plan a refactor at some point to address both of these -- ferret out numeric IDs and also move to consistent naming.

I'll drop in my bool.

Reply:
Looking at the code, I still think we need a standard numeric ID generator, not because it's a good idea for the database, but because a bunch of our tables already have this and it's good DRY practice. It will also help mark all the places that need to ID removed when the time comes to clean this up.

Reply:
Yes, because we need to remain backward compatible with web2py / DAL for the time being we will need to keep the numeric ids.

Reply:
the long term goal should be to clean up the schema as well!